File: languages/java/runtime/src/main/java/co/revaly/sdk/errors/RapFailureClass.java
package co.revaly.sdk.errors;

/**
 * The three typed failure classes of the RAP failover contract (docs/failover-contract.md §2). The
 * class — never the message text, latency, or wait heuristics — is what licenses a failover
 * decision.
 */
public enum RapFailureClass {
    /**
     * The request was received and rejected (400/401/403/404/422). Fix or decline — never fail
     * over: the same request fails anywhere.
     */
    PERMANENT_REJECTION,

    /**
     * The payment was definitively not processed (client-provable never-sent, or 503 with {@code
     * code: not_processed}). Safe to route to your own gateway immediately.
     */
    TRANSIENT_FAILURE,

    /**
     * The payment may have been processed (deadline after send, 5xx, ambiguous transport failure).
     * Reconcile before acting — failing over blind risks a double charge.
     */
    OUTCOME_UNKNOWN
}

File: languages/php/runtime/src/Errors/RapFailureClass.php
<?php

declare(strict_types=1);

namespace Revaly\Sdk\Errors;

/**
 * The three failure classes of failover-contract.md §2. The class — never the message
 * text, never latency — is what licenses (or forbids) failover.
 */
enum RapFailureClass: string
{
    /** Received and rejected. Fix or decline. Never fail over — the same request fails anywhere. */
    case PermanentRejection = 'PermanentRejection';

    /** Definitively not processed. Safe to route to your own gateway immediately. */
    case TransientFailure = 'TransientFailure';

    /** May have been processed. Reconcile before acting — failing over blind can double-charge. */
    case OutcomeUnknown = 'OutcomeUnknown';
}
